### PR TITLE
Theme switcher

### PR DESCRIPTION
**Pull request by kof**
https://github.com/webstudio-is/webstudio-designer/issues/108

Theme switch menu is behind a feature flag, to enable run `localStorage.feature="theme"`

### CODE REVIEW

File: apps/designer/app/routes/designer/$id.tsx [hunk 1]
```diff
@@ -1,13 +1,18 @@
...
 import { Designer, links } from "~/designer";
 import * as db from "~/shared/db";
 import config from "~/config";
 import env from "~/env.server";
+import { authenticator } from "~/services/auth.server";
 export { links };
 
-export const loader: LoaderFunction = async ({ params }) => {
+export const loader: LoaderFunction = async ({ request, params }) => {
+  const user = await authenticator.isAuthenticated(request);
```
Comment (kof) [inline apps/designer/app/routes/designer/$id.tsx:12]: this is a drive-by fix, we haven't checked in the designer if user is authenticated @SaraVieira

File: apps/designer/app/designer/shared/client-settings/settings.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
+import * as values from "./values";
```
Comment (kof) [inline apps/designer/app/designer/shared/client-settings/settings.ts:1]: is is currently unused, I built it before I realized we can't save theme setting into local storage, but I think we will have a plenty of use cases for local user settings in the local storage as an option to  user settings in the database

File: apps/designer/app/designer/features/topbar/menu/menu.tsx [hunk 3]
```diff
@@ -26,6 +34,41 @@ const menuItemCss = {
...
+            css={menuItemCss}
+            onSelect={() => {
+              setThemeSetting(setting);
+            }}
+          >
+            {labels[setting]}
+          </DropdownMenuCheckboxItem>
+        ))}
+      </DropdownMenuContent>
+    </DropdownMenu>
```
Comment (okonet) [inline apps/designer/app/designer/features/topbar/menu/menu.tsx:68]: Consider using `Select` as it should simplify the usage a lot.

File: apps/designer/app/designer/shared/client-settings/settings.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+const read = (): Settings => {
+  try {
+    const settings = localStorage.getItem(namespace);
+    if (settings === null) return {};
+    return JSON.parse(settings);
+  } catch (error) {
+    // We don't need to handle this one.
+  }
+  return {};
+};
```
Comment (okonet) [inline apps/designer/app/designer/shared/client-settings/settings.ts:19]: A test would be great. Also, just swallowing errors doesn't seem to be a good idea. Should those at least be logged and sent to Sentry?

File: apps/designer/app/shared/feature-flags/feature.ts [hunk 5]
```diff
@@ -28,5 +28,5 @@ export const isFeatureEnabled = (name: Name): boolean => {
   const defaultValue = flags[name];
   const envValue = parse(env.FEATURE).includes(name);
   const localValue = readLocal().includes(name);
-  return localValue ?? envValue ?? defaultValue;
+  return localValue || envValue || defaultValue;
```
Comment (okonet) [inline apps/designer/app/shared/feature-flags/feature.ts:31]: What is the difference?

File: apps/designer/app/shared/theme/theme.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  const query = matchMedia("(prefers-color-scheme: light)");
+  const queryColorScheme = (query: MediaQueryList | MediaQueryListEvent) =>
+    query.matches ? "light" : "dark";
+
+  system = queryColorScheme(query);
+
+  query.addEventListener("change", (queryEvent) => {
+    system = queryColorScheme(queryEvent);
+    renderThemeProps();
+  });
```
Comment (okonet) [inline apps/designer/app/shared/theme/theme.ts:24]: I could not find clean up code. Probably means there is a memory leak. Also, I think this should go into a hook.

File: apps/designer/app/designer/features/topbar/menu/menu.tsx (hunk 3 above)
Comment (kof) [reply apps/designer/app/designer/features/topbar/menu/menu.tsx:68]: That doesn't work, its an actual menu with submenus

File: apps/designer/app/shared/feature-flags/feature.ts (hunk 5 above)
Comment (kof) [reply apps/designer/app/shared/feature-flags/feature.ts:31]: includes `''`

File: apps/designer/app/shared/theme/theme.ts (hunk 6 above)
Comment (kof) [reply apps/designer/app/shared/theme/theme.ts:24]: this is a single event listened binding, its done once and never again

File: apps/designer/app/designer/shared/client-settings/settings.ts (hunk 4 above)
Comment (kof) [reply apps/designer/app/designer/shared/client-settings/settings.ts:19]: That test would be useless, because localStorage would be mocked or polyfilled and otherwise there is nothing in this function to test.

Comment (kof) [reply apps/designer/app/designer/shared/client-settings/settings.ts:19]: My strategy for tests is when there is complex logic or a bug already happened to as a regression test. We are not writing a library and testing every function is not viable or useful. Once we get to e2e tests, they will be able to cover real user flows.

Comment (okonet) [reply apps/designer/app/designer/shared/client-settings/settings.ts:19]: > This will only throw if basic localStorage is not available

It will also throw if JSON.parse won't work.

File: apps/designer/app/shared/feature-flags/feature.ts (hunk 5 above)
Comment (okonet) [reply apps/designer/app/shared/feature-flags/feature.ts:31]: Ouch. I wonder if there is a less ambiguous syntax or a helper function we could use for such cases.

File: apps/designer/app/shared/theme/theme.ts (hunk 6 above)
Comment (okonet) [reply apps/designer/app/shared/theme/theme.ts:24]: I see a function that can be called multiple times potentially :)

File: apps/designer/app/designer/features/topbar/menu/menu.tsx (hunk 3 above)
Comment (okonet) [reply apps/designer/app/designer/features/topbar/menu/menu.tsx:68]: Ah I see, it's a sub-menu in another menu.

**Review by okonet — APPROVED** [review]
I think it's fine for the current status of the project. We can refactor later.

File: apps/designer/app/designer/shared/client-settings/settings.ts (hunk 4 above)
Comment (kof) [reply apps/designer/app/designer/shared/client-settings/settings.ts:19]: right, but that can only mean the data in settings is corrupt, invalid json, what can we do about it except of ignoring it as if it doesn't exist?

Comment (okonet) [reply apps/designer/app/designer/shared/client-settings/settings.ts:19]: I guess reporting it to us so we know about this?

File: apps/designer/app/shared/feature-flags/feature.ts (hunk 5 above)
Comment (okonet) [reply apps/designer/app/shared/feature-flags/feature.ts:31]: Because it's confusing even to me (an experienced dev) :)

Comment (kof) [reply apps/designer/app/shared/feature-flags/feature.ts:31]: the reason I supported empty string is just because some may set that flag to en empty string in local storage to disable feature, its not really important, just thought would be more practical

Comment (kof) [reply apps/designer/app/shared/feature-flags/feature.ts:31]: I know remembered why I added those, its not because of empty string, its because of `false`
we return false from .includes, which makes the first value always take precedence with ??

File: apps/designer/app/designer/shared/client-settings/settings.ts (hunk 4 above)
Comment (kof) [reply apps/designer/app/designer/shared/client-settings/settings.ts]: done

File: apps/designer/app/shared/feature-flags/feature.ts (hunk 5 above)
Comment (okonet) [reply apps/designer/app/shared/feature-flags/feature.ts:33]: I get the logic but I think the amount of gotcha with `||` and `??` what makes me a bit uncomfortable reading it.

**Review by okonet — APPROVED** [review]
